### DISCUSSION TITLE
add Web Browser as a Player entity

### DISCUSSION THREAD

**Discussion by vk2him** [music-assistant/support#436] [Feature requests and ideas]
## [edit] NOTICE: This thread generally suggests using browser-mod as the solution for playing back music through the browser. This is no longer recommended. It is recommended to use the builtin web player although Snapcast with Snapweb is another option

The native Media browser has the ability to play media files on the actual device (iPad,iPhone,PC etc) that Home Assistant is being displayed on as can be seen in this screenshot:

![Image 006](https://user-images.githubusercontent.com/22878629/167739969-57e1c0a5-0f68-4583-89d9-911d35cb32c2.jpg)

It would be great if MA was also able to use the Web browser option

**Comment by marcelveldt** [comment]
Good suggestion.

TIP: You can already do that with this: https://github.com/thomasloven/hass-browser_mod
It has the option to use the browser as media_player

**Reply by paulius2k** [reply]
@marcelveldt 
> TIP: You can already do that with this: https://github.com/thomasloven/hass-browser_mod It has the option to use the browser as media_player

Could you guide me a little how I could do that?

Because when I include the browser_mod player into the Music Assistant configuration:
<img src="https://user-images.githubusercontent.com/24705262/181807788-a8a42ec0-be5a-41b2-839a-366a1446afff.png" width=50%>


my Music Assistant media player list becomes empty:
<img src="https://user-images.githubusercontent.com/24705262/181808056-7c986f71-a5e6-4733-9aff-2c1ea90fdd76.png" width=80%>

**Reply by marcelveldt** [reply]
@paulius2k make sure that it has a name.

**Reply by paulius2k** [reply]
@marcelveldt it has a name in my config.yaml:
```
browser_mod:
  prefix: "browser_mod_"
  devices:
    30da8608_5a769749:
      name: dell_latitude_chrome_device
```
in the integrations:
<img src="https://user-images.githubusercontent.com/24705262/181811824-07ed90f2-e0eb-4a45-b8a8-0bcc07eeda43.png" width=50%>


I even see that the `mass_*` entity is created and I can see it the `Media` application media players' list, and I can select it there and play media on it:
<img src="https://user-images.githubusercontent.com/24705262/181811448-0992287f-1891-46c0-894b-4a7352acaf98.png" width=50%>


But not on Music Assistant.

**Reply by vk2him** [reply]
> @marcelveldt it has a name in my config.yaml:
> 
> ```
> browser_mod:
>   prefix: "browser_mod_"
>   devices:
>     30da8608_5a769749:
>       name: dell_latitude_chrome_device
> ```
> 
> in the integrations: <img alt="" width="50%" src="https://user-images.githubusercontent.com/24705262/181811824-07ed90f2-e0eb-4a45-b8a8-0bcc07eeda43.png">
> 


@paulius2k - I saw this exact behaviour earlier this week where a browser_mod entity prevented any player being visible in the Players list.
After a lot of troubleshooting, I found I needed to add a friendly name to the browser_mod entity - even though it looked like it had one already I needed to force it as shown in this screenshot.

![Screen Shot 368](https://user-images.githubusercontent.com/22878629/181857865-b5258a30-a735-4da3-a559-91bf7f7bdc1b.jpg)

I suggest you click the pencil icon in the entity and typing a name there - this is what resolved it for me. Confusingly, I have other browser_mod entities that work fine without needing to do this, so this may not work for you.

Edit - note that in your config.yaml code example, the name is setting an alias which is different to a friendly name - I needed to set the friendly name for it to work.

**Reply by paulius2k** [reply]
@vk2him thank you!  I finally understood it :)  
In my case I needed to put a name on `media_player` entity, not the integration:
<img src="https://user-images.githubusercontent.com/24705262/181906421-29f65bd2-d479-4515-955d-e9e47b647d35.png" width=40%>

Now I see it in the list of media players in Music Assistant.

**Reply by ItRainsSmiles** [reply]
I wonder if this is already possible in MA 2 beta. I tried but failed.
Greetings :)

**Reply by OzGav** [reply]
Hi there. Not possible yet. Go and vote in the player provider poll

**Reply by ItRainsSmiles** [reply]
> Hi there. Not possible yet. Go and vote in the player provider poll

Already voted for "Web Browser" :) Thanks for the confirmation on this issue :)

**Reply by LordNex** [reply]
#I used to be able to use browser mod only music through a few years back. But it hasn't worked since I don't know how long. I don't see any of the music on my phone either

Not sure if there is a way but it would be nice if I could play all my Apple Music I've bought

**Reply by Rudd-O** [reply]
Browser mod is flaky and needs user interaction with the UI, it is also not embedded in the Music Assistant UI (you need a separate browser window) which automatically rules out playback on the same browser tab on mobile.

**Comment by erkr** [comment]
Moved to discussions, raise your votes!

**Comment by LordNex** [comment]
I totally agree with this. I was just out driving with my son and I wanted to play a track back on my NAS. I could queue it up to play but could only play to the google speakers I already have. It would be nice to be able to have my phone connected to my trucks Bluetooth and then be able to use MA as a music player while away from home. Would be such an awesome feature

**Comment by jdesai61** [comment]
browser_mod does help with this - but that means having to name each and every instance uniquely. What would be nice is being able to play to 'Current Browser' - the browser/app where I am using the HA app. The native Media player in Home Assistant has this capability and is typically the first option presented ("BROWSER") followed by a list of other media players.

**Reply by marcelveldt** [reply]
that is what this feature request is about... the browser mod is just a temporary solution until the final solution is implemented some day

**Reply by jdesai61** [reply]
> that is what this feature request is about... the browser mod is just a temporary solution until the final solution is implemented some day

Agree thanks. I somehow thought that browser_mod was being proposed as a solution.

**Comment by Clusters** [comment]
Unfortunately this "hack" does not work for me.

I have the browser_mod integration working and added it to Music Assistant. I can see the list and select the player like the other connected players
![grafik](https://user-images.githubusercontent.com/3765075/203520998-b062ef93-5845-4170-a7fa-2ac00e025c7f.png)

But if I select something in Music Assistant for replay it won't progress at all, nor do I hear any music. The browser_mod itself seem to work just fine, as I can select any (local) media to play on the media_player.my_cool_browser entity.
Only the mass equivalent of this entity can not play anything.

![grafik](https://user-images.githubusercontent.com/3765075/203522345-5ca7a75a-7ea0-42bb-bb3e-2dd1229388dc.png)

Anyone had a similiar issue with that?
Currently I have to select every single music file (one by one) for replay, as I can not get it working with Music Assistant.

the only relevant log entry:
`2022-11-23 13:30:03.311 WARNING (MainThread) [music_assistant.players.stream] Abort: client(s) did not connect within 10 seconds.`

edit: same behaviour with Chrome

**Reply by Clusters** [reply]
How does Music Assistant serve the local music? My issue might be, that I want to use a Web Browser outside of my local network (where the files are served via Samba to Home Assistant). I don't get it yet, why this shouldn't work (as the HASS Media Browser works without any hassle), but I got this finally running on a Firefox inside my local network.

**Reply by erkr** [reply]
Hi. You are right, not being on the same local network will be the issue. Music Assistant uses a custom stream port (TCP 8095 by default) to stream audio to players. Players must be able to reach the Home Assistant instance and this port.
Best Eric

**Reply by Clusters** [reply]
@erkr thanks for your answer. I should've read the README 🙈
I checked the port and it is indeed 8095 for my installation. I now have forwarded this port (TCP) to HASS, so it can be reached from the interwebs. However, the player still can't connect when trying to replay music on an outside webbrowser. 

The reason for that is, that the browser tries to access the media with the local IP address instead of the domain name of Home Assistant. I honestly don't know where this is coming from, as I have configured the domain name wherever it was possible. And ofc accessing it in the browser via its domain.
Can I control this somehow?

![grafik](https://user-images.githubusercontent.com/3765075/203954470-b44df2f3-6d82-4e81-8dff-7ef2bd6a147b.png)

**Reply by OzGav** [reply]
Marcel has responded before that what you asking for is not possible. I can’t find the exact detail at the moment but you are welcome to search the issues and discussions. Here is one response  he provided
_Music Assistant launches its own -internal only- webserver to serve the audio stream to players on the local network on TCP port 8095. The ip-address that is communicated to players is derived from whatever you fill in at the "internal URL" field of the HA config. So it only takes the IP-address from there (and even resolves dns if needed) and ignores everything else._

**Reply by Clusters** [reply]
@OzGav thanks for the hint. It took a bit, but I think I tracked down the discussion you were referring to https://github.com/music-assistant/hass-music-assistant/discussions/794

With that knowledge it was relatively easy to deploy the code changes that were necessary to get things working for me. As this behaviour is wanted by Marcel, I won't try to put that mainline, though ;-)

**Reply by jdesai61** [reply]
So does this already work now with 2.0 release? If so, I couldn't find how. If not, any progress? Thanks Music Assistant team and contributors for excellent project.

**Reply by marcelveldt** [reply]
No its not and will not make it into the 2.0 release (more likely 2.1). If you need a way to play music locally, use a local player like squeezelite. At some point we will introduce a builtin player for the webbrowser or even app but that is going to take a lot of time to implement the right way and not top priority at this point.

**Reply by madbrain76** [reply]
marcelveldt, glad to hear that's planned. It would seem beneficial if one is on the go and using headphones or car bluetooth, with the smartphone connected to VPN to access HA/MA. I'm still unsure if the voice assistants are up to the task of properly selecting content for the car case, but that's a separate issue.

**Comment by gaustabl** [comment]
Hello. I have read all the instructions, reinstalled browser mod, cleared cache and «everyting» else for the last week but i am not able to get browser as an alternative media player. I am about to give up, but maybe someone here has had the same problem and found the solution?   I am very thankful for any help, best regards Lars

**Reply by erkr** [reply]
Hi, I guess this discussion was working for MA v1 only. 
MA1 had a totally different architecture.  It was an integration in HA that used the existing music players. 
MA2 is an add-on with its own players (for various stability and maintenance reasons). As a consequence, it will not recognise additional players in HA.
Note: There is still an optional MA integration part. That integration currently allows to play media in HA. There is a planned feature to also publish HA players for use in the MA add-on. But to my knowledge it is planned beyond the official 1st release of MA 2.

**Reply by OzGav** [reply]
> to my knowledge it is planned beyond the official 1st release of MA 2

I just checked the backlog and it looks like Marcel might be working on this at the moment. So keep an eye out for future integration updates!

**Reply by gaustabl** [reply]
Thank for replyimg, then it makes sense that i annot get it to work😀

**Reply by SantiagoSotoC** [reply]
When snapcast player provider is released, using a browser as a player is possible with snapweb.
I guess it should be possible to integrate it into lovelace without any problem. I have no experience in custom lovelace development but I understand that javascript can be added.

**Comment by morpheus8888** [comment]
any news?

**Reply by OzGav** [reply]
Yes. Use Snapcast with the snapweb option and you can send music to any browser!

**Comment by alanmilinovic** [comment]
That is really a pity that you cannot use Web browser like in Media section of native HA.

**Reply by OzGav** [reply]
Well you can if you have a Snapcast server. We are looking at bundling a Snapcast server with MA.

**Reply by alanmilinovic** [reply]
That is nice to hear, but isn't that a bit overkill?

**Reply by OzGav** [reply]
Don't know you say that. If you would like to contribute the web browser provider that would be great.

**Reply by alanmilinovic** [reply]
All good, didn't meant anything bad.

**Reply by stu1811** [reply]
Any updates on this?

**Reply by OzGav** [reply]
What update do your want? Use the Snapcast provider

**Comment by angelnu** [comment]
Snapweb does not work when not home.

I usually connect to home-assistant exposed to the internet. I can do the same with MA (reverse proxy enforces user authentication) but this does not work with snapweb as it requires the client to expose a port which is not possible over the internet.

For time being I use a VPN connection - then snapweb works. Not as direct as playing in the browser but it works.

**Reply by stu1811** [reply]
Thanks for the explanation.

**Reply by mycroftsnm** [reply]
Does snapweb works well for you? half of the times i skip/pause a song the streams stop and i have to refresh and reconnect

**Comment by Rudd-O** [comment]
This would be swell.  I find it odd that I can't just stream from MA when I'm out of the home directly into my phone.  I can stream anywhere in the house, except where I am at.

**Comment by OzGav** [comment]
Please read the docs. Skip and pause not currently supported

**Comment by bibitocarlos** [comment]
Hope it comes one day ^^!

**Comment by bibitocarlos** [comment]
I saw that this request is considered as implemented, is it in the betas ?

**Reply by OzGav** [reply]
Read the whole discussion. Use Snapcast.

**Reply by alanmilinovic** [reply]
One day maybe, but not any time soon.

**Reply by bibitocarlos** [reply]
The title is "add web browser as a player entity", not "add snapcast as a player entity, which is a workaround and clearly not as efficient."
As such, it's not implemented.

**Reply by stu1811** [reply]
> Read the whole discussion. Use Snapcast.

Snapcast doesn't work remotely

**Reply by mycroftsnm** [reply]
And snapcast doesn't even allow pausing or changing track without breaking half of the time.

**Reply by SantiagoSotoC** [reply]
> Snapcast doesn't work remotely

snapcast does work remotely, but for security reasons it is not recommended, if you use a vpn like tailscale, you have no problem using it remotely.


> And snapcast doesn't even allow pausing or changing track without breaking half of the time.

Snapcast playback is going to improve, it takes time, I am not an expert in FFMPEG nor in python, but anyone is welcome to make a contribution, or create their own provider. You could also migrate snapweb to HA.

**Reply by alanmilinovic** [reply]
Just make a web browser player please, it is not something that doesn't exist already in HA. Snapcast is an overkill.

**Reply by SantiagoSotoC** [reply]
If you are so interested, you are more than welcome to program it yourself. If you have doubts about how MA works, write me on discord and I'll be happy to help. But here we are volunteers who do it in our free time, because we like it, in my case snapcast is the most suitable for my use case and that's why I maintain it.

**Reply by OzGav** [reply]
@alanmilinovic Contribute positively or not at all

**Reply by marcelveldt** [reply]
There are plans to add a builtin/web player but we have some other priorities first. As we are very limited in capacity all I'm asking for is a bit patience and keep it polite. Thanks!

**Comment by madbrain76** [comment]
Why not keep the RFE open until someone is able and willing to address it ? Tracking it doesn't mean implementing it right away, or even ever.
I have over 100 Mozilla bugs and RFEs that are still open, many filed as far back as 25 years ago when Mozilla started, because they are actually valid, but no one has allocated the resources yet in all that time. Perhaps no one ever will, or it will happen after I'm gone. Many of them have been resolved years after stopped working on the project, though, and thus keeping them open was the right thing to do. There is no need to take it personally when an issue isn't addressed or prioritized immediately.

When the issues/bugs/RFEs pile up, It helps ton to have a proper issue tracking system that lets you set things like priorities, component, severity, status, detailed resolution, version, OS, hardware, etc. It blows my mind that the default github issue tracker doesn't allow this. I hope they will in the future. In the meantime, the only thing that can be used seems to be tags.

**Comment by lscorcia** [comment]
To save some time for fellow enthusiasts, I think you should add a note to the documentation saying that there is currently no "built-in, secure and supported" way to play music from Music Assistant in a web browser when connected via the Internet. The suggested Snapcast + Snapweb solution only works when in the same network, or with a VPN.

**Comment by madbrain76** [comment]
One thing I can mention until this is implemented in MA is that BubbleUPNP seems to work for this purpose. On my Windows 10 PC, there is a free version I can run. It exposes a DLNA/UPNP render. You can use it in MA with the DLNA Player provider, and it will play through the PC's soundcard.

On my Android 14 phone, S22 Ultra, I purchased the BubbleUPNP app, which does the same thing, but with the phone's audio output. I had purchased it a long ago, and not specifically for use with MA. It is a $5.99 app. I have nor relationship with the author. I think it's a viable workaround until the feature is added to MA.

**Comment by Externalnet** [comment]
I cannot play music with the music assistant (not in my browser, and not on my phone with the ha app) it says no players. Removed the integration and add-on from ha. I will try in a year or so

**Comment by tylerlucas** [comment]
This should not be marked `implemented`, because the implementation requires additional software and a workaround.

**Reply by Rudd-O** [reply]
Agreed.  This is most definitely not implemented at all.

**Reply by OzGav** [reply]
Your opinions are noted

**Reply by marcelveldt** [reply]
I agree this should not be noted as implemented. Sure there are workarounds but streaming to the browser is not available today. All the workarounds that exist today are way too difficult as well imo. Lets keep the request open until we have some time to pick this up (not now, as we are swamped with other stuff)

**Reply by Rudd-O** [reply]
Congrats on all y'all getting to the finish line on this one!  Many a user will rejoice now.

**Comment by zeroskrull** [comment]
is there any tutorial about how to integrate broswer mod to this, everyone talk but no1 have a specific tutorial

**Reply by OzGav** [reply]
Did you read the bold text in the first post?

**Comment by ilar** [comment]
Right now the recommendation is to have Music Assistant in one window and snapcast web client in another tab.

It seems like an easy, MVP solution here would be to take advantage of the existing JS for the snapcast web client and have musicassistant spawn a web worker that runs the snapcast web client js. That way it survives navigating around homeassistant, but also brings it down to one tab that self-manages it's own snapcast client.

The code burden for support is GREATLY lessened as a result offloading all the complicated bits to snapcast, the only thing the webworker would need to manage is itself, and providing a framework for the snapcast web client to hook in to. No GUI would need to even be exposed from the snapcast PoV- all controls can be handled from MA.

**Comment by OzGav** [comment]
The first iteration of this is in 2.5.0b11

**Reply by vk2him** [reply]
Seems to be working well - on a Macbook and on an iPhone running the HA app - great work!!

**Reply by Dieghito72** [reply]
> The first iteration of this is in 2.5.0b11

Just gave it a try, absolutely amazing, I was missing this feature. Great work!

**Reply by fletchowns** [reply]
This is an awesome feature!! Nice work, thank you for implementing. It has now become my primary music player on desktop :)

**Comment by Externalnet** [comment]
I am not understanding this. I Installed Music Assitant as an addon in Home Assitant, also added the HA integration, added my NAS SMB share with audio files in de MA webui, and added the Home Assitant MediaPlayers and Home Assitant Plugin Provider. I can see my audio files in MA, now how do I play a file from my Home Assistant app or from a browser ? I see an audio file in de MA weggui and when I whant to play it says no player selected (the list is empty) I cannot figure out how to play an audio file. I also do not see the SMB share under Media in Home Assistant. It also would not play/stream to my Shield Android TV. I removed eveything again because this is not working for me

**Reply by OzGav** [reply]
Sounds like you haven’t selected any entities when you added the HA media player provider. Please review the docs.

**Reply by Externalnet** [reply]
I added Shield Android TV but it was not playing any audio. MA was connected because I could shutdown Shield Android TV through MA. Is there any way I can play audio through the Home Assitant interface ? (Web, companion app) It would be awesome if I could play my local audio files through the HA Android companion app.

**Reply by OzGav** [reply]
Yes you can ply locally but you need to be running the beta version of the server or wait  a little while until 2.5 comes out.

**Reply by Externalnet** [reply]
Thank you very much for the info, I did not run the beta version, I will try again.

**Comment by OzGav** [comment]
The feature request has been implemented successfully. As such this feature request will be closed and locked. If there are enhancements desired then open a new [feature request](https://github.com/orgs/music-assistant/discussions/categories/feature-requests-and-ideas). If there is a question about how this works then check the [documentation](https://beta.music-assistant.io/), or ask a question in the [Q&A](https://github.com/orgs/music-assistant/discussions/categories/q-a) area or on Discord. If there are problems then ask a question in the [Q&A](https://github.com/orgs/music-assistant/discussions/categories/q-a) area or on Discord or open an [issue](https://github.com/music-assistant/support/issues). Thanks for your interest in seeing this idea being brought to life.
